File: spec/go/ax25sdl/sdl_test.go
package ax25sdl

import "testing"

// TestStatePagesHaveTransitions asserts every generated state-machine
// page declared a non-empty Transitions slice. A page with no
// transitions is almost certainly a codegen bug — the validator
// rejects YAML pages with zero transitions before they reach the
// emitter.
func TestStatePagesHaveTransitions(t *testing.T) {
	pages := []StatePage{
		DataLinkAwaitingConnection,
		DataLinkAwaitingV22Connection,
		DataLinkAwaitingRelease,
		DataLinkConnected,
		DataLinkDisconnected,
	}
	for _, p := range pages {
		if len(p.Transitions) == 0 {
			t.Errorf("%s/%s has no transitions", p.Machine, p.State)
		}
	}
}

// TestSubroutinesPageHasBodies asserts the figc4.7 page declared its
// subroutines. The spec figure has 13; the current graphml has authoring
// bugs on Establish_Data_Link and Establish_Extended_Data_Link (n50
// SABM is missing its outgoing edge), so the tool emits 11 — those two
// are skipped with a warning. When the graphmls are redrawn this bumps
// back to 13.
func TestSubroutinesPageHasBodies(t *testing.T) {
	const expected = 11
	if got := len(DataLinkSubroutines.Subroutines); got != expected {
		t.Errorf("expected %d subroutines on figc4.7, got %d", expected, got)
	}
}

// TestActionKindStringRoundTrips asserts every ActionKind value
// produces a non-"unknown" String, catching a generated file that
// names an out-of-range kind constant.
func TestActionKindStringRoundTrips(t *testing.T) {
	for k := SignalUpper; k <= InternalOut; k++ {
		if k.String() == "unknown" {
			t.Errorf("ActionKind(%d) → unknown", k)
		}
	}
}

File: spec/ts/src/ax25sdl/sdl.test.ts
import { describe, it, expect } from "vitest";

import {
  DataLinkAwaitingConnection,
  DataLinkAwaitingV22Connection,
  DataLinkAwaitingRelease,
  DataLinkConnected,
  DataLinkDisconnected,
  DataLinkSubroutines,
} from "./index.js";
import type { StatePage, ActionKind } from "./index.js";

// State pages must declare at least one transition. The codegen
// validator rejects YAML pages with zero transitions before they
// reach the emitter, so an empty page is a generator bug.
describe("state pages", () => {
  const pages: readonly StatePage[] = [
    DataLinkAwaitingConnection,
    DataLinkAwaitingV22Connection,
    DataLinkAwaitingRelease,
    DataLinkConnected,
    DataLinkDisconnected,
  ];

  for (const p of pages) {
    it(`${p.machine}/${p.state} has transitions`, () => {
      expect(p.transitions.length).toBeGreaterThan(0);
    });
  }
});

// figc4.7 declares thirteen subroutines in the spec. The graphml edge
// fix for #11 (SABM out-edge from n50) restored Establish_Data_Link
// and Establish_Extended_Data_Link, so all 13 transcribe.
describe("figc4.7 subroutines", () => {
  it("has 13 entries", () => {
    expect(DataLinkSubroutines.subroutines).toHaveLength(13);
  });

  it("UI_Check appears", () => {
    const names = DataLinkSubroutines.subroutines.map((s) => s.name);
    expect(names).toContain("UI_Check");
  });
});

// ActionKind is a string-literal union; spot-check that generated
// pages use values from the declared set. The compile-time check is
// already enforced by the type system; this is belt-and-braces.
describe("action kinds", () => {
  const allowed = new Set<ActionKind>([
    "signal_upper", "signal_lower", "processing", "subroutine", "internal_out",
  ]);

  it("every generated verb has a known kind", () => {
    for (const p of [DataLinkAwaitingConnection, DataLinkConnected, DataLinkDisconnected]) {
      for (const t of p.transitions) {
        for (const a of t.actions) {
          expect(allowed.has(a.kind)).toBe(true);
        }
      }
    }
  });
});
